### ISSUE TITLE
Output image quality issues in RGB image generation tasks

### ISSUE THREAD

**Issue by sheep1222** [IBM/aihwkit#699]
I want to use this tool to simulate, train, and infer with a convolutional neural network for RGB image generation tasks, which incorporates an attention mechanism. I have tried using the convert_to_nalog function and various RPU configurations, but the results were not satisfactory. So, I would like to ask, which configuration you would recommend without considering the non-idealities of the devices, or is there a method to directly map the trained parameters for inference?

**Comment by PabloCarmona** [comment]
Hello @sheep1222, and thanks for trying the kit! Can you please elaborate more on your setup with some code examples of what you are trying to achieve?

**Comment by sheep1222** [comment]
Sorry, I've been busy lately and haven't replied in a timely manner.
My program is like this：
import torch
import torch.nn as nn
import torchvision.utils as vutils
import torch.optim as optim
import torchvision
from torchvision import datasets, transforms
from torch.utils.data import DataLoader
import matplotlib.pyplot as plt
import numpy as np
import imageio
from aihwkit.nn import AnalogLinear, AnalogSequential
from aihwkit.optim import AnalogSGD
from aihwkit.nn.conversion import convert_to_analog
from aihwkit.simulator.presets import MixedPrecisionEcRamMOPreset
from aihwkit.simulator.rpu_base import cuda
from aihwkit.simulator.configs import MappingParameter

mapping = MappingParameter(weight_scaling_omega=0.8)
RPU_CONFIG = MixedPrecisionEcRamMOPreset(mapping=mapping)

# Hyperparameters
batch_size = 128
image_size = 64
nc = 3
nz = 100
ngf = 64
ndf = 64
num_epochs = 5
lr = 0.0002
beta1 = 0.5

USE_CUDA = 0
if cuda.is_compiled():
    USE_CUDA = 1
device = torch.device("cuda" if USE_CUDA else "cpu")

# Data Preparation
transform = transforms.Compose([
    transforms.Resize(image_size),
    transforms.CenterCrop(image_size),
    transforms.ToTensor(),
    transforms.Normalize([0.5] * nc, [0.5] * nc)
])

dataset = datasets.ImageFolder(root="./data/", transform=transform)
dataloader = DataLoader(dataset, batch_size=batch_size, shuffle=True, num_workers=4)


# Generator
class Generator(nn.Module):
    def __init__(self):
        super(Generator, self).__init__()
        self.main = nn.Sequential(
            nn.ConvTranspose2d(nz, ngf * 8, 4, 1, 0, bias=False),
            nn.BatchNorm2d(ngf * 8),
            nn.ReLU(True),

            nn.ConvTranspose2d(ngf * 8, ngf * 4, 4, 2, 1, bias=False),
            nn.BatchNorm2d(ngf * 4),
            nn.ReLU(True),

            nn.ConvTranspose2d(ngf * 4, ngf * 2, 4, 2, 1, bias=False),
            nn.BatchNorm2d(ngf * 2),
            nn.ReLU(True),

            nn.ConvTranspose2d(ngf * 2, ngf, 4, 2, 1, bias=False),
            nn.BatchNorm2d(ngf),
            nn.ReLU(True),

            nn.ConvTranspose2d(ngf, nc, 4, 2, 1, bias=False),
            nn.Tanh()
        )

    def forward(self, x):
        return self.main(x)


generator = Generator()
generator = convert_to_analog(generator, RPU_CONFIG)
generator = generator.to(device)
gen_opt = AnalogSGD(generator.parameters(), lr=2e-2)
gen_opt.regroup_param_groups(generator)

# Discriminator
class Discriminator(nn.Module):
    def __init__(self):
        super(Discriminator, self).__init__()
        self.main = nn.Sequential(
            nn.Conv2d(nc, ndf, 4, 2, 1, bias=False),
            nn.LeakyReLU(0.2, inplace=True),

            nn.Conv2d(ndf, ndf * 2, 4, 2, 1, bias=False),
            nn.BatchNorm2d(ndf * 2),
            nn.LeakyReLU(0.2, inplace=True),

            nn.Conv2d(ndf * 2, ndf * 4, 4, 2, 1, bias=False),
            nn.BatchNorm2d(ndf * 4),
            nn.LeakyReLU(0.2, inplace=True),

            nn.Conv2d(ndf * 4, ndf * 8, 4, 2, 1, bias=False),
            nn.BatchNorm2d(ndf * 8),
            nn.LeakyReLU(0.2, inplace=True),

            nn.Conv2d(ndf * 8, 1, 4, 1, 0, bias=False),
            nn.Sigmoid()
        )

    def forward(self, x):
        return self.main(x).view(-1, 1).squeeze(1)


discriminator = Discriminator()
discriminator = convert_to_analog(discriminator, RPU_CONFIG)
discriminator = discriminator.to(device)
disc_opt = AnalogSGD(discriminator.parameters(), lr=2e-2)
disc_opt.regroup_param_groups(discriminator)

# Loss and Optimizers
criterion = nn.BCEWithLogitsLoss()

# Training Loop
G_losses = []
D_losses = []
image_list = []

fixed_noise = torch.randn(batch_size, nz, 1, 1, device=device)

for epoch in range(num_epochs):
    for i, (data, _) in enumerate(dataloader):
        real_images = data.to(device)
        batch_size = real_images.size(0)

        real_labels = torch.full((batch_size,), 1.0, device=device)
        fake_labels = torch.full((batch_size,), 0.0, device=device)

        # Train Discriminator
        disc_opt.zero_grad()
        output_real = discriminator(real_images)
        loss_real = criterion(output_real, real_labels)

        noise = torch.randn(batch_size, nz, 1, 1, device=device)
        fake_images = generator(noise)
        output_fake = discriminator(fake_images.detach())
        loss_fake = criterion(output_fake, fake_labels)

        loss_D = (loss_real + loss_fake) / 2
        loss_D.backward()
        disc_opt.step()

        # Train Generator
        gen_opt.zero_grad()
        output_fake = discriminator(fake_images)
        loss_G = criterion(output_fake, real_labels)
        loss_G.backward()
        gen_opt.step()

        if i % 100 == 0:
            print(
                f"Epoch [{epoch + 1}/{num_epochs}] Step [{i}/{len(dataloader)}] Loss_D: {loss_D.item():.4f} Loss_G: {loss_G.item():.4f}")

        G_losses.append(loss_G.item())
        D_losses.append(loss_D.item())

    # Save generated images
    with torch.no_grad():
        fake_images = generator(fixed_noise).detach().cpu()
        image_list.append(fake_images)

# Save GIF
images = []
for epoch_images in image_list:
    # Convert to [0, 255] and uint8
    grid = np.transpose(vutils.make_grid(epoch_images[:64], padding=2, normalize=True), (1, 2, 0))
    grid = (grid * 255).to(torch.uint8).numpy()  # Scale to [0, 255] and convert to uint8
    images.append(grid)
imageio.mimsave('dcgan.gif', images, fps=2)

# Visualize final generated images
plt.figure(figsize=(8, 8))
plt.axis("off")
plt.title("Generated Images")
grid = np.transpose(torchvision.utils.make_grid(fake_images[:64], padding=2, normalize=True), (1, 2, 0))
plt.imshow(grid.numpy())
plt.show()

I also try FloatingPointDevices and ConstantStepDevice, both are default configurations. The result is not good.

**Comment by kaoutar55** [comment]
Hello @sheep1222 are you trying to train with no noise? Can you explain what you are trying to do. Simulate training or simulate inference in analog?

**Comment by sheep1222** [comment]
Hello,I am trying to simulate both train and inference in analog.In order to achieve a good result, I tried simulating an ideal noise free device, but the effect was still not good.Have you tried applying this kit to color image generation or segmentation？

**Comment by PabloCarmona** [comment]
Please @sheep1222, take a look at our new feature from @jubueche, hope that helps you out with the thing you're trying to achieve!
